### PR TITLE
Compute Neighbor during Map Inclusion Verification

### PR DESCRIPTION
**Pull request by gdbelvin**
Have the Map verification algorithm compute the proper neighbor nodes. This is a requirement of verification correctness. 

This updates and simplifies the `Sibling()` logic in `storage.NodeID`. 
The only functional change is masking off the unused bits.

### CODE REVIEW

File: merkle/common.go [hunk 1]
```diff
@@ -32,3 +32,37 @@ func bit(index []byte, i int) uint {
...
+func flipBit(index []byte, i int) []byte {
+	r := make([]byte, len(index))
+	copy(r, index)
+	IndexBits := len(index) * 8
+	bIndex := (IndexBits - i - 1) / 8
+	r[bIndex] ^= 1 << uint(i%8)
+	return r
+}
+
+// Neighbor returns index with only the left i bits set and the i'th bit flipped.
```
Comment (Martin2112) [inline merkle/common.go]: Can you give a bit more details on this? In particular we have a Sibling() function for logs, which could lead to confusion.

File: integration/maptest/map_test.go [hunk 2]
```diff
@@ -88,7 +88,7 @@ func verifyGetMapLeavesResponse(getResp *trillian.GetMapLeavesResponse, indexes
 		leafHash := incl.GetLeaf().GetLeafHash()
 		proof := incl.GetInclusion()
 
-		if got, want := leafHash, hasher.HashLeaf(treeID, index, hasher.BitLen(), leaf); !bytes.Equal(got, want) {
+		if got, want := leafHash, hasher.HashLeaf(treeID, index, 0, leaf); !bytes.Equal(got, want) {
```
Comment (Martin2112) [inline integration/maptest/map_test.go]: Do we need tests at non zero depth?

Comment (gdbelvin) [inline integration/maptest/map_test.go]: The map doesn't support leaves at non-zero depths, so I think not.

Comment (Martin2112) [inline integration/maptest/map_test.go]: Do we need that parameter then?

File: merkle/common.go (hunk 1 above)
Comment (gdbelvin) [inline merkle/common.go]: I'll move this functionality to `Sibling`

File: integration/maptest/map_test.go (hunk 2 above)
Comment (gdbelvin) [inline integration/maptest/map_test.go]: Right now, strictly speaking, we don't. 
If we were to support leaves at non-zero levels in the future, we would.

Comment (Martin2112) [inline integration/maptest/map_test.go]: Supporting that would be a fairly big change. I could argue for removing this parameter until it's used.

File: storage/types.go [hunk 3]
```diff
@@ -282,20 +282,59 @@ func (n *NodeID) CoordString() string {
...
+		r[depthBytes-1] = r[depthBytes-1] & leftmask[depth%8]
+	}
+	if depth < n.PrefixLenBits {
+		n.PrefixLenBits = depth
+	}
+	n.Path = r
+	return n
+}
+
+// Neighbor sets this node to be it's own neighbor.
```
Comment (Martin2112) [inline storage/types.go]: Still not sure if this means the same as Sibling(), or is it equivalent to Sibling() for maps. Can you expand the documentation?

File: integration/maptest/map_test.go (hunk 2 above)
Comment (gdbelvin) [inline integration/maptest/map_test.go]: Sounds good

Comment (gdbelvin) [inline integration/maptest/map_test.go]: Here's the PR to remove it: #769

File: storage/types.go (hunk 3 above)
Comment (gdbelvin) [inline storage/types.go]: This is doing exactly the same thing as the previous version of `Sibling`. The steps have just been broken up for clarity. 

`Neighbor` is doing the following snippet from the previous: 

```
bi = n.PathLenBits() - n.PrefixLenBits + i ==
height = n.PathLenBits() - (n.PrefixLenBits - i)
r[i].SetBit(bi, n.Bit(bi)^1)
```

Mask Left is doing this snippet
```
r[i].PrefixLenBits =  n.PrefixLenBits - i
```

Comment (Martin2112) [inline storage/types.go]: OK. it looks like this is only used by maps / SMT. In the C++ version we had Sibling() which was used by logs but that wasn't carried over to the new code. I wanted to make sure that there wasn't going to be API confusion.

Comment (gdbelvin) [inline storage/types.go]: Sounds good.  In either case, the original functionality has been maintained.

**Comment by Martin2112** [comment]
Is this all merged now via other PRs?

**Comment by gdbelvin** [comment]
This is not merged yet.  Several other PRs have been rebased on it though.
